docs/source/daml/code-snippets/Keys.daml

```diff
@@ -94,177 +96,251 @@ template Delegation
       do
         fetchByKey @Keyed lookupKey
 
-lookupTest = scenario do
+template Helper
```

Comment:
What's the idea of this Helper contract?

Reply:
Context: 

> 
> The `submit` and `submitMustFail` functions are actually methods of the `HasSubmit` class.
> 
> ```haskell
> class HasSubmit m cmds where
>     submit         : HasCallStack => Party -> cmds a -> m a
>     submitMustFail : HasCallStack => Party -> cmds a -> m ()
> ```
> 
> And we have two relevant instances,
> 
> ```haskell
> instance HasSubmit Scenario Update
> instance HasSubmit Script   Commands
> ```
> 

Now, the important part is that `Commands` isn't quite as powerful (on its own) as `Update` - there are a bunch of functions with `Update a` return type that don't have an equivalent `Commands a` function. Crucially, `Update` has a `Monad`/`Action` instance, while `Commands` is only `Applicative`. 

For instance, there are a few tests where we do things like

```
-- Maintainer can fetch
  submit sig do
    (cid, keyed) <- fetchByKey @Keyed sig
    assert (keyedCid == cid)
```

When translating this to `Script` + `Commands`, we notice two things: 

1. There's no `Commands` equivalent to `fetchByKey` - though as you mentioned earlier we do have the somewhat similar `queryContractId : (Template t, IsParties p, HasCallStack) => p -> ContractId t -> Script (Optional t)`
2. The `do` block isn't going to work in a `Commands` context, since the second line (`assert ...`) depends  on the result of the first one (`(cid, keyed) <- ...`)

In this particular case, we can rewrite the test somewhat straightforwardly, by replacing `fetchByKey` with `queryContractId` and moving the `assertion` out of the `Update/Commands` block up to the `Scenario/Script` context:

```
-- Maintainer can fetch
  Some keyed <- queryContractId sig sig
  assert (keyedCid == cid)
```

However, this isn't always possible. In particular, if the original test had `submitMustFail` instead of `submit`, there's no way we can get the result of the `Update/Commands` block in the `Scenario/Script` context (since `submitMustFail : HasCallStack => Party -> cmds a -> m ()`)

The advice I got from @cocreature on this (and which I've since noticed in a few places of our codebase, as a sort of folk knowledge) was to move the `Update` body to a choice in a `Helper` template. This way, the `Update` itself needs no change at all. To _use_ the body, however, we need to summon a `Helper` contract and exercise the choice, but that's one of the things we _can_ do in `Commands` with `createAndExerciseCmd : (Template t, Choice t c r) => t -> c -> Commands r)`.


TL;DR: for some tests, we'd want to have a function `: Update a -> Commands a`. The closest thing we can have is physically moving the `Update a` code into a choice `C` of a helper template `T`, and then call `createAndExerciseCmd T C`.

Reply:
cool, that makes sense.